src/System.ComponentModel.Composition/src/Microsoft/Internal/GenerationServices.cs

```diff
@@ -168,10 +171,25 @@ public static void LoadValue(this ILGenerator ilGenerator, object value)
         /// <returns></returns>
         public static void AddItemToLocalDictionary(this ILGenerator ilGenerator, LocalBuilder dictionary, object key, object value)
         {
-            Assumes.NotNull(ilGenerator);
-            Assumes.NotNull(dictionary);
-            Assumes.NotNull(key);
-            Assumes.NotNull(value);
+            if(ilGenerator == null)
+            {
+                throw new ArgumentNullException(nameof(ilGenerator));
```

Comment:
@ViktorHofer Shouldn't we throw NullReferenceException for arguments qualified with `this` for extension methods?

Reply:
It seems like this code is not really optimized. ilGenerator is never used as the extension method is used in only one branch where ilGenerator is never null: https://source.dot.net/#System.ComponentModel.Composition/System/ComponentModel/Composition/MetadataViewGenerator.cs,186

I would replace the whole check with `Debug.Assert(ilGenerator != null);`. You could also do that at other places after you checked that the this argument is never null.